### DISCUSSION TITLE
WasmScript Proposal

### DISCUSSION THREAD

**Discussion by ashtonmeuser** [ashtonmeuser/godot-wasm#50] [Proposals]
Stemmed from https://github.com/ashtonmeuser/godot-wasm/issues/45 and several discussions with the [V-Sekai](https://github.com/V-Sekai) folks.

The motivation for WasmScript is to be able to easily attach a Wasm module to a Godot [Node](https://docs.godotengine.org/en/stable/classes/class_node.html). As such, WasmScript will inherit [Script](https://docs.godotengine.org/en/stable/classes/class_script.html) (or, more accurately, `Ref<Script>`). It should have an associated Wasm module (likely a path string) that will be instantiated (via [`Wasm.load()`](https://github.com/ashtonmeuser/godot-wasm/wiki/Class-Documentation:-Wasm#error-load--packedbytearray-bytecode-dictionary-imports-)) when the Node's `_init()` or `_ready()` methods are called.

Because Wasm uses only integer and float primitives, exposing Godot's API will involve quite a bit of marshalling to and from module memory. All integer values will likely need to be treated as pointers to a memory offset (including those representing integers within Godot).

WasmScript will, by default, provide a very limited slice of Godot API to the Wasm module via function imports. The following functions will be imported by default.
- `get(property_offset: i64, property_length: i64)` exposes [`Object.get()`](https://docs.godotengine.org/en/stable/classes/class_object.html#id1)
- `set(property_offset: i64, property_length: i64, value_offset: i64)` exposes [`Object.set()`](https://docs.godotengine.org/en/stable/classes/class_object.html#id6)
- `call(function_offset: i64, function_length: i64, args_offset: i64)` exposes [`Object.call()`](https://docs.godotengine.org/en/stable/classes/class_object.html#class-object-method-call)

The above API can be greatly simplified by a library used during Wasm module creation. Such a library could be provided for common languages targeting Wasm e.g. AssemblyScript, Rust.

Some questions:
1. What exactly is contained in a WasmScript file? Is it JSON or another data interchange format? At minimum, the WasmScript instance must specify the Wasm module that it is to load. How is this done?
2. How can the exposed Godot API be expanded upon? For example, a project e.g. V-Sekai would like to expose a `change_color()` method to the Wasm module via a `VSWasmScript` class. How is this done?

**Comment by ashtonmeuser** [comment]
Expanding on question 1. The contents of `GDScript` is GDScript, contents of `CSharpScript` is C#, contents of [`ECMAScript`](https://github.com/Geequlim/ECMAScript) is JavaScript. What are the contents of `WasmScript`? Is it some config that points to a Wasm module and defines imports? Is it literal Wasm bytecode?

**Reply by fire** [reply]
I thought about this. 

`Ref<Script>` is a Resource which we means we get properties and get and set. Like the .tres format.

We can store the buffer of the wasm as a property.

**Reply by ashtonmeuser** [reply]
> We can store the buffer of the wasm as a property.

How does this look from a user's perspective? They have a .wasm file and want to assign it to a Node. Keep in mind bytecode is only needed transiently to compile so duplicating into some Resource property (in addition to .wasm file) may be a little wasteful.

**Reply by fire** [reply]
The standard godot system is except for like text files, and like raw pngs, we wrap external imports into a resource (copying into a resource.)

**Reply by ashtonmeuser** [reply]
Okay interesting. Thanks for the info. So with that assumption, how do you see this working? Do the resources load/copy the Wasm bytecode at runtime? How does the actual WasmScript resource define which file to pull from?

**Reply by fire** [reply]
I slept on this.

How difficult is it to embed an entire language into wasm?

1. input gdscript text
2. run gdscript as a wasm module

3. input c++ folder
4. get a wasm compiled in wasm

**Reply by fire** [reply]
The godot resource is either defined to be exactly the wasm binary or it can be a godot resource and a binary buffer inside that contains the wasm. 

You may want to store input variables in the resource and metadata for input types.

**Reply by fire** [reply]
gdscript ".gd" is defined as exactly the binary buffer (of the string) as the resource.
